test/distributed_graph_generator_test_utils.h

```diff
@@ -0,0 +1,560 @@
+		}
+
+		template <typename BufferT>
+		buffer_access_step reduce(BufferT& buf, const bool include_current_buffer_value);
+
+		// FIXME: Misnomer (not a "buffer access")
+		template <typename HostObjT>
+		buffer_access_step affect(HostObjT& host_obj, experimental::side_effect_order order = experimental::side_effect_order::sequential) {
+			return chain<buffer_access_step>([&host_obj, order](handler& cgh) { host_obj.add_side_effect(cgh, order); });
+		}
+	};
+
+  public:
+	template <typename Name, int Dims>
+	buffer_access_step device_compute(const range<Dims>& global_size, const id<Dims>& global_offset) {
+		std::deque<action> actions;
+		actions.push_front([global_size, global_offset](handler& cgh) { cgh.parallel_for<Name>(global_size, global_offset, [](id<Dims>) {}); });
+		return buffer_access_step(m_dctx, std::move(actions));
+	}
+
+	template <typename Name, int Dims>
+	buffer_access_step device_compute(const nd_range<Dims>& execution_range) {
+		std::deque<action> actions;
+		actions.push_front([execution_range](handler& cgh) { cgh.parallel_for<Name>(execution_range, [](nd_item<Dims>) {}); });
+		return buffer_access_step(m_dctx, std::move(actions));
+	}
+
+	template <int Dims>
+	buffer_access_step host_task(const range<Dims>& global_size) {
+		std::deque<action> actions;
+		actions.push_front([global_size](handler& cgh) { cgh.host_task(global_size, [](partition<Dims>) {}); });
+		return buffer_access_step(m_dctx, std::move(actions));
+	}
+
+	buffer_access_step master_node_host_task() {
+		std::deque<action> actions;
+		actions.push_front([](handler& cgh) { cgh.host_task(on_master_node, [] {}); });
+		return buffer_access_step(m_dctx, std::move(actions));
+	}
+
+	buffer_access_step collective_host_task(experimental::collective_group group) {
+		std::deque<action> actions;
+		actions.push_front([group](handler& cgh) { cgh.host_task(experimental::collective(group), [](const experimental::collective_partition&) {}); });
+		return buffer_access_step(m_dctx, std::move(actions));
+	}
+
+  private:
+	dist_cdag_test_context& m_dctx;
+
+	task_builder(dist_cdag_test_context& dctx) : m_dctx(dctx) {}
+};
+
+class command_query {
+	friend struct command_query_testspy;
+	friend class dist_cdag_test_context;
+
+	class query_exception : public std::runtime_error {
+		using std::runtime_error::runtime_error;
+	};
+
+  public:
+	// TODO Other ideas:
+	// - remove(filters...)						=> Remove all commands matching the filters
+	// - executes(global_size)					=> Check that commands execute a given global size (exactly? at least? ...)
+	// - writes(buffer_id, subrange) 			=> Check that commands write a given buffer subrange (exactly? at least? ...)
+	// - find_one(filters...)					=> Throws if result set contains more than 1 (per node?)
+
+	template <typename... Filters>
+	command_query find_all(Filters... filters) const {
+		assert_not_empty(__FUNCTION__);
```

Comment:
What does that assertion actually do?

Reply:
Calling for example `query_a.have_type(command_type::execution)` doesn't make much sense if `query_a` does not contain any commands. What should it return, true or false? I decided it'd be best to just throw in that case. For `find_all` it's actually somewhat debatable whether it should be allowed to further "narrow down" an already empty set of commands (maybe in some generic context it would make sense), but I decided to err on the side of caution for now.

Reply:
Ah, this is a member function. I assumed this would do something akin to `assert(__FUNCTION__ != "")`